### PR TITLE
[WIP] Transformer Layers

### PR DESCRIPTION
**Pull request by cgarciae**
Hey! 

# What does this PR do?
Adds the `TransformerEncoderLayer`, `TransformerEncoder`, `TransformerDecoderLayer`, `TransformerDecoder`, and `Transformer` modules, their APIs and implementations are ported as faithful as possible from PyTorch (with obvious deviations when needed).

## Tasks
- [ ] Implementation
- [ ] Tests
- [ ] Docs
- [ ] Example

## Comments
Posting PR early to get some feedback, starting with if this addition is actually of interest to Flax, naming conventions, style, implementation, etc.

### CODE REVIEW

**Comment by jheek** [comment]
Hi, we have so far refrained from adding Transformer to the main nn library.
The main purpose of the nn library is to provide the primitive building blocks of neural nets but not specific models.
Of course this is somewhat ambiguous and (Self)Attention for example was considered a "grey area" which eventually was added (but later simplified).
For new additions to the nn library it's best to first open a feature request issue so we can discuss whether it's in scope.

**Comment by cgarciae** [comment]
I see. That makes sense. I'll leave it here for now in case someone wants to provide free feedback and will find it a home later.

**Comment by avital** [comment]
Hi @cgarciae -- just doing some GitHub bookkeeping now. Sounds like we won't merge this, so I'll close the PR but hopefully others can find it if they search (if you want to edit the issue with the relevant keywords for what this enables that would help people find it)

**Comment by DBraun** [comment]
Thanks for this code. I'm starting to use it. I noticed that these two lines should be swapped:

https://github.com/cgarciae/flax/blob/a4dc414385d1ef8365933f0ca4414c728ef25fcd/flax/nn/transformer.py#L248-L249

I also had to do some things with passing deterministic to Dropout layers. Could @cgarciae update it since you know it much better? It would be helpful to see it in https://github.com/google/flax/tree/main/flax/nnx/examples (although I haven't transitioned to nnx yet either)
